src/lib/message/ExchangeContext.cpp

```diff
@@ -0,0 +1,1664 @@
+}
+#endif // CHIP_CONFIG_ENABLE_RELIABLE_MESSAGING
+
+/**
+ * @overload
+ */
+CHIP_ERROR ExchangeContext::HandleMessage(ChipMessageInfo * msgInfo, const ChipExchangeHeader * exchHeader, PacketBuffer * msgBuf)
+{
+    return HandleMessage(msgInfo, exchHeader, msgBuf, NULL);
+}
+
+/**
+ * Handle a message in the context of an Exchange. This method processes ACKs and duplicate messages
+ * and then invokes the application handler.
+ *
+ * Note on OnMessageReceived and the umhandler argument:
+ * When the ExchangeManager creates a new EC for an inbound message,
+ * OnMessageReceived is set by default to a handler that drops the message, so
+ * any future message on the exchange is discarded unless the application
+ * installs an OnMessageReceived handler.
+ * The unsolicited message that triggers the creation of the EC is
+ * handled by an UMH, which is passed to this method via the umhandler param.
+ *
+ * @param[in] msgInfo       General CHIP message information for the incoming message.
+ * @param[in] exchHeader    CHIP exchange information for the incoming message.
+ * @param[in] msgBuf        PacketBuffer containing the payload of the incoming message.
+ * @param[in] umhandler     Pointer to a message receive callback function; if this function
+ *                          is not NULL it will be used in place of the OnMessageReceived function
+ *                          installed in the ExchangeContext.
+ *
+ */
+CHIP_ERROR ExchangeContext::HandleMessage(ChipMessageInfo * msgInfo, const ChipExchangeHeader * exchHeader, PacketBuffer * msgBuf,
+                                          ExchangeContext::MessageReceiveFunct umhandler)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+#if CHIP_CONFIG_ENABLE_RELIABLE_MESSAGING
+    const uint8_t * p        = NULL;
+    uint32_t PauseTimeMillis = 0;
+#endif
+
+    // We hold a reference to the ExchangeContext here to
+    // guard against Close() calls(decrementing the reference
+    // count) by the application before the CHIP Exchange
+    // layer has completed its work on the ExchangeContext.
+    AddRef();
+
+    if (msgInfo->MessageVersion == kChipMessageVersion_V2)
+    {
+#if CHIP_CONFIG_ENABLE_RELIABLE_MESSAGING
+        if (exchHeader->Flags & kChipExchangeFlag_AckId)
+        {
+            err = RMPHandleRcvdAck(exchHeader, msgInfo);
+        }
+        if (exchHeader->Flags & kChipExchangeFlag_NeedsAck)
+        {
+            // Set the flag in message header indicating an ack requested by peer;
+            msgInfo->Flags |= kChipMessageFlag_PeerRequestedAck;
+
+            // Set the flag in the exchange context indicating an ack requested;
+            SetPeerRequestedAck(true);
+
+            if (!ShouldDropAck())
+            {
+                err = RMPHandleNeedsAck(msgInfo);
+            }
+        }
+#endif
+    }
+
+    // If the message is a duplicate and duplicates are not allowed for this exchange.
+    if ((msgInfo->Flags & kChipMessageFlag_DuplicateMessage) && !AllowDuplicateMsgs)
```

Comment:
Not quite sure why CHIP would support both modes where messages and duplicated, and not... (even cryptographically). Why wouldn't we just not allow duplicates, and drop?

Reply:
I think this was only relevant for the Trickle protocol; should be elided from the code